### PR TITLE
[TRANSFORMATIONS] Create python binding for pattern::Optional

### PR DESCRIPTION
**Pull request by CuriousPanCake**
[TRANSFORMATIONS] Create python binding for pattern::Optional

Expose the C++ op::pattern::Optional to Python in order to
simplify patterns creation.
Cover the functionality with the dedicated tests.

### Tickets:
CVS-133523

Signed-off-by: Andrii Staikov <andrii.staikov@intel.com>

### CODE REVIEW

File: src/bindings/python/tests/test_transformations/test_pattern_ops.py [hunk 1]
```diff
@@ -84,6 +84,93 @@ def test_any_input_predicate():
     assert matcher.match(param)
     assert not matcher.match(slope)
 
+def test_optional_full_match():
```
Comment (akuporos) [inline src/bindings/python/tests/test_transformations/test_pattern_ops.py:87]:
```suggestion

def test_optional_full_match():
```
to fix linter

Comment (akuporos) [reply src/bindings/python/tests/test_transformations/test_pattern_ops.py:87]: Please, add extra line before each test to fix linter job

File: src/core/src/pattern/op/optional.cpp [hunk 2]
```diff
@@ -43,6 +43,7 @@ bool ov::pass::pattern::op::Optional::match_value(Matcher* matcher,
     // Turn the Optional node into WrapType node to create a case where the Optional node is present
     ov::OutputVector input_values_to_optional = input_values();
     size_t num_input_values_to_optional = input_values_to_optional.size();
+    bool same_type = pattern_value.get_element_type() == graph_value.get_element_type();
```
Comment (iefode) [inline src/core/src/pattern/op/optional.cpp:46]: Did not get the usage of `type` in non-related type pattern.. So, could you please explain?

Comment (CuriousPanCake) [reply src/core/src/pattern/op/optional.cpp:46]: Imagine you're matching 2 nodes like that. Please, pay attention that the pattern has no inputs.
image

With the current implementation of `if (matcher->match_value(pattern, graph_value) || num_input_values_to_optional == 0) {`
even if `matcher->match_value()` fails (as it should in this case), we will still hit the case of `num_input_values_to_optional == 0` resulting in the nodes matching which is wrong. That's why I suggest checking for a number of inputs to optional being zero **_only_** if  the types of nodes match.

File: src/bindings/python/src/pyopenvino/graph/passes/pattern_ops.cpp [hunk 3]
```diff
@@ -482,6 +483,110 @@ static void reg_pattern_any_input(py::module m) {
...
+                      :param input: input node
+                      :type input: openvino.runtime.Node
+    )");
+
+    optional_type.def(py::init([](const std::vector<std::string>& type_names, const std::shared_ptr<ov::Node>& input, const Predicate& pred) {
+                          return std::make_shared<ov::pass::pattern::op::Optional>(get_types(type_names), input, pred);
+                      }),
+                      py::arg("type_names"),
+                      py::arg("input"),
+                      py::arg("pred"),
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/passes/pattern_ops.cpp:571]:
```suggestion
                      py::arg("predicate"),
```

File: src/bindings/python/src/pyopenvino/graph/passes/pattern_ops.cpp [hunk 4]
```diff
@@ -482,6 +483,110 @@ static void reg_pattern_any_input(py::module m) {
...
+
+                      :param type_names: node type. For example: ["opset8.Abs", "opset8.Relu"]
+                      :type type_names: List[str]
+    )");
+
+    optional_type.def(py::init([](const std::vector<std::string>& type_names, const Predicate& pred) {
+                          return std::make_shared<ov::pass::pattern::op::Optional>(get_types(type_names), pred);
+                      }),
+                      py::arg("type_names"),
+                      py::arg("pred"),
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/passes/pattern_ops.cpp:506]:
```suggestion
                      py::arg("predicate"),
```

File: src/bindings/python/src/pyopenvino/graph/passes/pattern_ops.cpp [hunk 5]
```diff
@@ -482,6 +483,110 @@ static void reg_pattern_any_input(py::module m) {
...
+                      :param predicate: Function that performs additional checks for matching.
+                      :type predicate: function
+    )");
+
+    optional_type.def(py::init([](const std::vector<std::string>& type_names, const ov::Output<ov::Node>& input, const Predicate& pred) {
+                          return std::make_shared<ov::pass::pattern::op::Optional>(get_types(type_names), input, pred);
+                      }),
+                      py::arg("type_names"),
+                      py::arg("input"),
+                      py::arg("pred"),
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/passes/pattern_ops.cpp:522]:
```suggestion
                      py::arg("predicate"),
```

Comment (CuriousPanCake) [reply src/bindings/python/src/pyopenvino/graph/passes/pattern_ops.cpp]: Fixed.

File: src/bindings/python/src/pyopenvino/graph/passes/pattern_ops.cpp (hunk 4 above)
Comment (CuriousPanCake) [reply src/bindings/python/src/pyopenvino/graph/passes/pattern_ops.cpp]: Fixed.

File: src/bindings/python/src/pyopenvino/graph/passes/pattern_ops.cpp (hunk 3 above)
Comment (CuriousPanCake) [reply src/bindings/python/src/pyopenvino/graph/passes/pattern_ops.cpp:578]: Fixed.

File: src/common/snippets/src/op/subgraph.cpp [hunk 6]
```diff
@@ -257,16 +257,16 @@ auto Subgraph::wrap_node_as_subgraph(const std::shared_ptr<ov::Node>& node) -> s
 }
 
 void Subgraph::fill_empty_output_names(const Output<Node>& target_output_node, const Output<Node>& replacement_output_node) {
-    OPENVINO_SUPPRESS_DEPRECATED_START
+    //OPENVINO_SUPPRESS_DEPRECATED_START
```
Comment (itikhono) [inline src/common/snippets/src/op/subgraph.cpp:260]: I believe this a part of another PR

File: src/common/transformations/tests/control_flow/unroll_tensor_iterator_test.cpp [hunk 7]
```diff
@@ -510,7 +510,7 @@ void collect_legacy_tensor_names(const std::shared_ptr<ov::Model>& model, std::v
 // after SeqToTI transformation:  Params -> TI -> Unsqueeze -> Results
 // after UnrollTI transformation: Params -> [Unrolled TI] -> Unsqueeze -> Results
 // No new tensor names after [UnrolledTI]
-TEST(TransformationTests, CheckTensorNamesAfterConvertToTIAndUnrolling) {
+TEST(TransformationTests, DISABLED_CheckTensorNamesAfterConvertToTIAndUnrolling) {
```
Comment (itikhono) [inline src/common/transformations/tests/control_flow/unroll_tensor_iterator_test.cpp:513]: I believe this a part of another PR

File: src/core/tests/pattern.cpp [hunk 8]
```diff
@@ -510,37 +510,35 @@ TEST(pattern, matching_optional) {
                          std::make_shared<op::v0::Abs>(c)));
 }
 
-TEST(pattern, optional_full_match) {
+TEST(pattern, DISABLED_optional_full_match) {
```
Comment (itikhono) [inline src/core/tests/pattern.cpp:513]: it's better to use some skip list file and mention the jira ticket

Comment (itikhono) [reply src/core/tests/pattern.cpp:513]: As I can see, these tests were not configured to work with a skip list files
Ok, lets leave it as it is, but please add a comment to each test about the issue description and a jira ticket

File: src/common/snippets/src/op/subgraph.cpp (hunk 6 above)
Comment (CuriousPanCake) [reply src/common/snippets/src/op/subgraph.cpp]: Yeah, I think, it was late at night :)
Fixed.

File: src/common/transformations/tests/control_flow/unroll_tensor_iterator_test.cpp (hunk 7 above)
Comment (CuriousPanCake) [reply src/common/transformations/tests/control_flow/unroll_tensor_iterator_test.cpp]: Fixed

File: src/bindings/python/tests/test_transformations/test_pattern_ops.py [hunk 9]
```diff
@@ -85,6 +86,104 @@ def test_any_input_predicate():
     assert not matcher.match(slope)
 
 
+@pytest.mark.skip("Optional is not working properly yet CVS_133523")
+def test_optional_full_match():
```
Comment (itikhono) [inline src/bindings/python/tests/test_transformations/test_pattern_ops.py:90]: Do we have any tests that are not skipped?
I believe that some of them should pass.

**Review by itikhono — APPROVED** [review]
in general LGTM

**Review by jiwaszki — APPROVED** [review]
LGTM, please replace `CVS_133523` in skip marks later or link the another ticket in it.

File: src/core/src/pattern/op/optional.cpp (hunk 2 above)
Comment (iefode) [reply src/core/src/pattern/op/optional.cpp]: Anyway, a bit strange solution in mix with previous PR. I will check it myself in my branch.

Comment (CuriousPanCake) [reply src/core/src/pattern/op/optional.cpp]: We have removed it as it's not really working. But the case it a valid scenario in which the current implementation is failing.

**Comment by CuriousPanCake** [comment]
> LGTM, please replace `CVS_133523` in skip marks later or link the another ticket in it.

Fixed.

**Pull request merged**